designs/2022-languages/README.md

```diff
@@ -0,0 +1,723 @@
+- Repo: eslint/eslint
+- Start Date: 2022-11-25
+- RFC PR: https://github.com/eslint/rfcs/pull/99
+- Authors: Nicholas C. Zakas
+
+# ESLint Language Plugins
+
+## Summary
+
+This proposal provides a way for plugin authors to define parsers, rules, etc., for any arbitrary programming languages. This would allow ESLint to work on any type of programming language for which there is a plugin defined, not just for JavaScript and its variants.
+
+## Motivation
+
+Over the nine years ESLint has been in existence, it has grown into being not just the most popular JavaScript linter, but arguably the most popular linter for any programming language. Twitter is filled with people lamenting that they don't have an "ESLint for X", where X is some other file type they're working with. Numerous projects have popped up that mimic a lot of what ESLint does at it's core:
+
+* [Stylelint](https://stylelint.io/);
```

Comment:
Personal opinion as a Stylelint maintainer (not representing Stylelint team opinion).

Appreciate the thought that was put in this RFC. Fantastic work!

I think for new linters such possibility to focus on rules (and the bridge between language and ESLint core) and don't think about the core could really help to get traction fast and have less maintenance.

For Stylelint I'm not convinced yet that transforming into ESLint plugin is a good decision.

Stylelint team would need to write and maintain completely new thing — plugin for ESLint, to bridge the gap between ESLint core and Stylelint rules.

Change all Stylelint rules to work with ESLint. Adopt to new parameters passed to each rule. Change how violation is reported. Change how auto-fixing work (from AST modification to text-based fixer).

Lose all of Stylelint plugins. Stylelint plugins authors would need to create ESLint version of their plugin.

User would have to navigate in all of this, change their configurations, editor settings, packages, CI.

For now I see low Return On Investment if Stylelint team and ecosystem would be doing this change. It would take a lot of time to change for Stylelint repositories. Many Stylelint plugins would be dead, because authors won't write the same plugin for ESLint.

---

Maybe there is alternative solution, where we don't try to hook everything into ESLint. Instead ESLint provides a package with linter core, that could be used as a dependency for any other linter. Common thing like finding and parsing configs, plugin system, how rules are ran, etc. But there is always a risk that such package won't be used by anyone...

Reply:
I was thinking of this more in terms of the last thing you mentioned: ESLint as the core of a linter upon which other linters can be built. I didn't mean to imply that Stylelint should be rewritten in the ESLint format -- that would be too much. I was more thinking that that Stylelint itself might actually be able to be reimplemented as an ESLint plugin while maintaining support for the existing rules and plugins. (This also might not be possible, but was at least what I was thinking.)

At the end of the day, we also want the ESLint core to be generic enough that anyone can build any type of linter they want on top of it.